iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -137,99 +154,320 @@ public static RegistryManager getRegistyManager(String iotHubConnectionString) t
+                }
+
+                SSLContext sslContext = SSLContextBuilder.buildSSLContext(IntegrationTest.x509CertificateGenerator.getPublicCertificate(), IntegrationTest.x509CertificateGenerator.getPrivateKey());
+                DeviceClient client = new DeviceClient(registryManager.getDeviceConnectionString(testDeviceIdentity.getDevice()), protocol, sslContext);
+                testDeviceIdentity.setDeviceClient(client);
+                return testDeviceIdentity;
+            }
         }
     }
 
-    public static TestModuleIdentity getTestModule(String iotHubConnectionString, IotHubClientProtocol protocol, AuthenticationType authenticationType) throws URISyntaxException, IOException, IotHubException, InterruptedException, ModuleClientException, GeneralSecurityException
+    /**
+     * Return a module identity and client that can be used for a test. If a recycled identity is available to use, this method will prioritize using
+     * that instead of creating a new identity. When creating a new identity, this method will not create a batch of identities
+     * since there is no bulk add service API for just modules. This method will also prioritize
+     * returning cached identities with twin changes over cached identities with no twin changes, but needCleanTwin can be used
+     * to never return a cached identity with twin changes.
+     *
+     * For instance, a test method that just wants to send telemetry from a module identity should set needCleanTwin to false
+     * so that it can use any available module. Conversely, a test method that involves setting reported or desired properties
+     * should set needCleanTwin to true to avoid previous twin state interfering with the upcoming test.
+     *
+     * @param iotHubConnectionString The connection string for the IoT Hub where the identity will be registered to.
+     * @param protocol The device side protocol for the client to use.
+     * @param authenticationType The device side authentication type for the client to use.
+     * @param needCleanTwin True if the returned module identity needs to not have any pre-existing desired or reported properties, false otherwise.
+     * @return A {@link TestDeviceIdentity} that was either recycled from a previous test when possible, or was just created for this test.
+     * @throws URISyntaxException If the connection string cannot be parsed.
+     * @throws IOException If the registry addition of a module fails.
+     * @throws IotHubException If the registry addition of a module fails.
+     * @throws GeneralSecurityException If creating the x509 certificates for an x509 module fails.
+     */
+    public static TestModuleIdentity getTestModule(String iotHubConnectionString, IotHubClientProtocol protocol, AuthenticationType authenticationType, boolean needCleanTwin) throws URISyntaxException, IOException, IotHubException, InterruptedException, ModuleClientException, GeneralSecurityException
```

Comment:
Similar to getTestDevice, except there is no batch add API for modules, so instead of adding 100 at a time, it just creates the identities at the time that they are requested without caching any for future callers. Modules only get cached when a test finished up with them.